openwisp_controller/subnet_division/handlers.py

```diff
@@ -0,0 +1,111 @@
+# exists in this queue. If it exists changes are made to related objects.
+
+
+@receiver(
+    pre_save,
+    sender=SubnetDivisionRule,
+    dispatch_uid='subnet_division_rule_changed_fields',
+)
+def subnet_division_rule_changed_fields(sender, instance, **kwargs):
+    try:
+        db_instance = SubnetDivisionRule.objects.get(id=instance.id)
+    except SubnetDivisionRule.DoesNotExist:
+        # This rule does not exists in database.
+        # No operation is needed to be performed.
+        return
+    else:
+        # Check which fields of instance is modified
+        # NOTE: Open-ended implementation to allow change in all
+        # fields of SubnetDivisionRule in future.
+        # Currently only changing label and number of IPs is allowed.
+        modified_fields = {}
+        for field in db_instance._meta.fields:
+            instance_value = getattr(instance, field.name)
+            db_value = getattr(db_instance, field.name)
+            if instance_value != db_value:
+                modified_fields[field.name] = db_value
+
+        _subnet_division_rule_update_queue[str(instance.id)] = modified_fields
+
+
+@receiver(
+    post_save,
+    sender=SubnetDivisionRule,
+    dispatch_uid='update_subnet_division_rule_related',
+)
+def update_subnet_division_rule_related(sender, instance, created, **kwargs):
+    def update_subnet_division_index_entries():
+        # Update related objects appropriately.
+        # NOTE: Currently only changing label and number of IPs is implemented/allowed.
+        try:
+            modified_fields = _subnet_division_rule_update_queue.pop(str(instance.id))
+        except KeyError:
+            return
+        else:
+            if 'label' in modified_fields:
+                tasks.update_subnet_division_index.delay(rule_id=str(instance.id))
+                tasks.update_subnet_name_description(rule_id=str(instance.id))
+            if 'number_of_ips' in modified_fields:
+                tasks.provision_extra_ips.delay(
+                    rule_id=str(instance.id),
+                    old_number_of_ips=modified_fields['number_of_ips'],
+                )
+
+    if created:
+        return
+    transaction.on_commit(update_subnet_division_index_entries)
+
+
+@receiver(
+    post_delete, sender=SubnetDivisionRule, dispatch_uid='delete_provisioned_subnets'
+)
+def delete_provisioned_subnets(instance, **kwargs):
+    # Deleting an object of SubnetDivisionRule will automatically delete
+    # it's related SubnetDivisionIndex objects due to "on_delete=CASCADE".
+    # Similarly, deleting a Subnet object will automatically delete IpAddress
+    # objects related to it.
+    # Hence, this method only executes query of deleting Subnets provisioned by
+    # this corresponding SubnetDivisionRule.
+    def delete_subnets():
+        Subnet.objects.filter(
+            organization_id=instance.organization_id, name__startswith=instance.label
+        ).delete()
```

Comment:
same here, I'd still make a model method for this because it could be overridden with custom logic